text/0000-stack-buildpacks.md

```diff
@@ -0,0 +1,299 @@
+# Meta
+[meta]: #meta
+- Name: Stack Buildpacks
+- Start Date: 2020-08-27
+- Author(s): [@jkutner](@jkutner)
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: N/A
+
+# Summary
+[summary]: #summary
+
+This is proposal for a new type of buildpack that runs against a stack in order to extend it in ways that are only possible by running privileged commands.
+
+# Motivation
+[motivation]: #motivation
+
+Normally, buildpacks do not run as root. This is an intentional design decision that ensures operations like `rebase` will work on day-2.
+
+However, many applications and buildpacks require modifications to the stack they run on, such as adding system packages or custom certificates. For this reason, we need a mechanism that buildpack authors and buildpack users and leverage to extend their stacks.
+
+# What it is
+[what-it-is]: #what-it-is
+
+- *root buildpack* - a new type of buildpack that runs as the root user
+- *stack buildpack* - a type of root buildpack that runs against the stack image(s) instead of an app. It is distinguished by a static list of mixins it can provide.
+- *userspace buildpack* - the traditional definition of a buildpack (i.e. does not run as root, and runs against an app)
+
+A new type of buildpack, called a Stack buildpack, may run against a stack (both build and run images) in order to extend it in ways that are only possible by running privileged commands. A stackpack may also define a list of mixins that it provides to the stack. In this way, a stack that is missing a mixin required by a buildpack may have that mixin provided by a stack buildpack.
+
+A stack provider may choose to include stack buildpacks with the stack they distribute. If a stack includes any stack buildpacks, the following will occur when the build phase starts:
+
+1. If any requested mixin is not provided by the stack, the lifecycle will compare the missing mixins to the static list of mixins provided by stack buildpacks. If any mixins are still not provided, the build will fail.
+1. The lifecycle will run the detect phase for all stackpacks defined in the builder.
+1. The lifecycle will execute the stack buildpack build phase for all passing stackpack(s). If no stack buildpacks pass detect, the build will continue the build phase as normal (running userspace buildpacks).
+1. After the lifecycle's build phase, the lifecycle will begin a new phase called _extend_.
+1. During the extend phase, stack buildpacks that passed detection will run against the run images accordingly (see details below).
+
+
+# How it Works
+[how-it-works]: #how-it-works
+
+ A stack buildpack (a.k.a. stackpack) is a special case of buildpack that has the following properties:
+
+* Is run as the `root` user
+* Configured with `privileged = true` in the `buildpack.toml`
+* Can only create one layer
+* Does not have access to the application source code
+* Is run before all regular buildpacks
```

Comment:
I thought it ran during the extend phase (according to line 38) which occurs after all regular buildpacks

Reply:
Or perhaps the extend phase is only for the run image being updated. Ok nvm